tensorflow/compiler/xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -219,49 +203,226 @@ static StatusOr<absl::optional<se::blas::AlgorithmType>> DoUncachedGemmAutotune(
+    if (init_cublas_data) {
+      InitializeBuffer(stream, op->shape().element_type(), &rng_state, buffer);
+    }
+    return buffer;
+  };
 
-  tensorflow::mutex_lock cache_lock(autotune_cache_mu);
-  auto it = autotune_cache.find(key);
-  int64_t autotuning_requests = cache_hits + cache_misses;
-  if (autotuning_requests && autotuning_requests % 10 == 0) {
-    VLOG(2) << "Autotuning cache hits/(hits + misses): " << cache_hits << "/"
-            << autotuning_requests;
-  }
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase lhs_buffer,
+                      get_initialized_buffer(instr->operand(0)));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase rhs_buffer,
+                      get_initialized_buffer(instr->operand(1)));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase output_buffer,
+                      get_initialized_buffer(instr));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase reference_result_buffer,
+                      get_initialized_buffer(instr));
+
+  int64_t batch_size = gemm_config.batch_size();
+#if GOOGLE_CUDA && CUDA_VERSION >= 11000
+  GpuGemmConfig config = GetGpuGemmConfig(instr);
+  const Shape& output_shape = config.output_shape;
+  std::unordered_set<PrimitiveType> enabled_types = {F16, F32, F64, C64, C128};
+
+  if (config.use_cublaslt &&
+      enabled_types.find(output_shape.element_type()) != enabled_types.end()) {
+    se::blas::MatrixDescriptor lhs_matrix, rhs_matrix, output_matrix;
+    std::tie(lhs_matrix, rhs_matrix, output_matrix) =
+        PopulateInputOutputMatrices(config, lhs_buffer, rhs_buffer,
+                                    output_buffer);
+
+    DCHECK(output_matrix.transpose == se::blas::Transpose::kNoTranspose);
+    tensorflow::DataType dtype;
+    se::blas::DataType blas_dtype;
+    switch (output_shape.element_type()) {
+      case F16:
+        dtype = tensorflow::DataTypeToEnum<Eigen::half>::value;
+        blas_dtype = se::blas::ToDataType<Eigen::half>::value;
+        break;
+      case F32:
+        dtype = tensorflow::DataTypeToEnum<float>::value;
+        blas_dtype = se::blas::ToDataType<float>::value;
+        break;
+      case F64:
+        dtype = tensorflow::DataTypeToEnum<double>::value;
+        blas_dtype = se::blas::ToDataType<double>::value;
+        break;
+      case C64:
+        dtype = tensorflow::DataTypeToEnum<complex64>::value;
+        blas_dtype = se::blas::ToDataType<complex64>::value;
+        break;
+      case C128:
+        dtype = tensorflow::DataTypeToEnum<complex128>::value;
+        blas_dtype = se::blas::ToDataType<complex128>::value;
+        break;
+    }
 
-  if (it != autotune_cache.end()) {
-    cache_hits++;
-    VLOG(4) << "Autotuning cache hit, using algorithm: "
-            << (it->second.has_value() ? absl::StrCat(*(it->second))
-                                       : "<generic>");
-    return it->second;
-  }
-  cache_misses++;
-  VLOG(4) << "Autotuning cache miss";
+    bool allow_tf32 = tensorflow::tensor_float_32_execution_enabled();
```

Comment:
`tensor_float_32_execution_enabled` detail should normally IMO be pushed into SE level. @reedwm what do you think? I don't see it used elsewhere in XLA.

Reply:
Agreed this should probably be checked in SE if this code is using SE, for consistency. Currently, TF32 is only checked within SE and tests.

I didn't read through this PR though so I could be missing important context.